### PR TITLE
add gfl_trt

### PR DESCRIPTION
**Pull request by Richard-mei**
add `GFocalHead` support in mmdeploy.   
rewrite `GFLHead.get_boxes`，`GFLHead.forward_single`，and replacing `F.linear` with `F.conv2d` in `Integral` module.

### CODE REVIEW

File: mmdeploy/codebase/mmdet/models/dense_heads/base_dense_head.py [hunk 1]
```diff
@@ -10,6 +10,212 @@
 from mmdeploy.utils import Backend, get_backend, is_dynamic_shape
 
 
+@FUNCTION_REWRITER.register_rewriter(
```
Comment (VVsssssk) [inline mmdeploy/codebase/mmdet/models/dense_heads/base_dense_head.py:13]: Hi, thanks for your PR. You can create a new file such as gfl_head.py in this dir and add your modification. And then, could you add the unit test for this function in tests/test_codebase/test_mmdet/test_mmdet_models.py?

Comment (Richard-mei) [reply mmdeploy/codebase/mmdet/models/dense_heads/base_dense_head.py:13]: Hi, I rewritte my changes to the gfl_head.py file. It can be tested in the same way as a general conversion to tensorrt.

Comment (grimoire) [reply mmdeploy/codebase/mmdet/models/dense_heads/base_dense_head.py:13]: Since the code has been moved to new files. The code in base_dense_head can be removed.

File: mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    cls_feat = x
+    reg_feat = x
+    for cls_conv in self.cls_convs:
+        cls_feat = cls_conv(cls_feat)
+    for reg_conv in self.reg_convs:
+        reg_feat = reg_conv(reg_feat)
+    cls_score = self.gfl_cls(cls_feat)
+    bbox_pred = scale(self.gfl_reg(reg_feat)).float()  # B, C(4*17), H, W
+    batch_preds = []
+    for b_pred in bbox_pred:
```
Comment (grimoire) [inline mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py:206]: Split a batched tensor with for loop is not a good idea. TensorRT can not support dynamic batch on such structure.

File: mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        batch_bboxes,
+        batch_scores,
+        max_output_boxes_per_class,
+        iou_threshold=iou_threshold,
+        score_threshold=score_threshold,
+        pre_top_k=pre_top_k,
+        keep_top_k=keep_top_k)
+
+
+def integral(x, reg_max=16, project=None):
```
Comment (grimoire) [inline mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py:175]: It would be better to write a integral with batch support. for example:

```python

    def batched_integral(self, intergral, x):
        batch_size = x.size(0)
        x = F.softmax(x.reshape(batch_size, -1, intergral.reg_max + 1), dim=2)
        x = F.linear(x,
                     intergral.project.type_as(x).unsqueeze(0)).reshape(
                         batch_size, -1, 4)
        return x
```

File: mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py (hunk 2 above)
Comment (grimoire) [reply mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py:206]: And since integral is performed in get_bboxes in mmdetection, I think we can write a batched integral and use it in gfl_head__get_bbox, rewrite of forward is not necessary.

**Comment by grimoire** [comment]
It would be a good idea to add a unit test in https://github.com/open-mmlab/mmdeploy/blob/master/tests/test_codebase/test_mmdet/test_mmdet_models.py .

File: mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py (hunk 3 above)
Comment (Richard-mei) [reply mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py:175]: Thank you so much. I modified and tested the code as you suggested and it works fine. But the shape of `bbox_pred` obtained by `forward_single` is B, C, H, W. I tried to execute `bbox_pred.permute(0,2,3,1)` in `forward_single` and `get_bbox` respectively, and test the speed, the former is faster.

Comment (Richard-mei) [reply mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py:175]: Your suggestion helped me a lot, however, I'm not familiar with mmdeploy's uint test, so I'm sorry I can't give more help.  
I am able to provide the command that runs when I test.  
> - deploy:  
```python tools/deploy.py configs/mmdet/detection/detection_tensorrt_dynamic-320x320-1344x1344.py ../MMDetection/configs/gfl/gfl_r50_fpn_1x_coco.py ../MMDetection/checkpoints/gfl_r50_fpn_1x_coco_20200629_121244-25944287.pth ../MMDetection/demo/demo.jpg --work-dir test_work/test_GFL/ --show --device cuda:0```    
> - test:  
```python tools/test.py configs/mmdet/detection/detection_tensorrt_dynamic-320x320-1344x1344.py ../MMDetection/configs/gfl/gfl_r50_fpn_1x_coco.py --model test_work/test_GFL/end2end.engine --speed-test --device cpu```

Comment (grimoire) [reply mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py]: > Thank you so much. I modified and tested the code as you suggested and it works fine. But the shape of `bbox_pred` obtained by `forward_single` is B, C, H, W. I tried to execute `bbox_pred.permute(0,2,3,1)` in `forward_single` and `get_bbox` respectively, and test the speed, the former is faster.

Thanks for the test. How faster if we use nonbatched integral? I think it is worth supporting batch inference with a little performance.

**Comment by grimoire** [comment]
Please fix the lint. You can run `pre-commit run --all-file` to do all the check.

@VVsssssk please provide some help about unit test.

**Comment by Richard-mei** [comment]
> Please fix the lint. You can run `pre-commit run --all-file` to do all the check.
>   
```flake8...................................................................Passed
seed isort known_third_party.............................................Passed
isort....................................................................Passed
yapf.....................................................................Passed
Trim Trailing Whitespace.................................................Passed
Check Yaml...............................................................Passed
Fix End of Files.........................................................Passed
Fix requirements.txt.....................................................Passed
Fix double quoted strings................................................Passed
Check for merge conflicts................................................Passed
Fix python encoding pragma...............................................Passed
Mixed line ending........................................................Passed
Markdownlint.............................................................Passed
codespell................................................................Passed
docformatter.............................................................Passed  
```

File: mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py [hunk 4]
```diff
@@ -0,0 +1,203 @@
+import torch
```
Comment (RunningLeon) [inline mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py:1]: could add copyright at top line
```
# Copyright (c) OpenMMLab. All rights reserved.
```

Comment (Richard-mei) [reply mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py:1]: Hi, do I need to resubmit to add copyright:`# Copyright (c) OpenMMLab. All rights reserved.`?
Also, I have a little doubt about the speed test. I have performed the speed test of the model under tensorrt on cuda and cpu respectively. Why is the FPS based on cpu very high, much higher than the speed based on cuda? Only the `--device cpu/cuda` parameter was changed during testing. My laptop cpu is r9-5900hs and gpu is 3060.**But I found that `nvidia-smi` shows gpu usage in both cases.**    
cpu:  `python tools/test.py configs/mmdet/detection/detection_tensorrt_dynamic-320x320-1344x1344.py ../MMDetection/configs/gfl/gfl_r50_fpn_1x_coco.py --model test_work/test_GFL/end2end.engine --speed-test --device cpu`  
Screenshot_select-area_20220210191630  
cuda:  `python tools/test.py configs/mmdet/detection/detection_tensorrt_dynamic-320x320-1344x1344.py ../MMDetection/configs/gfl/gfl_r50_fpn_1x_coco.py --model test_work/test_GFL/end2end.engine --speed-test --device cuda`  
Screenshot_select-area_20220210191906

Comment (grimoire) [reply mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py:1]: @RunningLeon

File: mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+    Rewrite this function to deploy model, (by richard.)
+    """
+    cls_feat = x
+    reg_feat = x
+    for cls_conv in self.cls_convs:
+        cls_feat = cls_conv(cls_feat)
+    for reg_conv in self.reg_convs:
+        reg_feat = reg_conv(reg_feat)
+    cls_score = self.gfl_cls(cls_feat)
+    bbox_pred = scale(self.gfl_reg(reg_feat)).float().permute(
```
Comment (grimoire) [inline mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py:201]: Could permute in `gfl_head__get_bbox` before `batched_integral`. So we can remove `gfl_head__forward_single`.

Comment (grimoire) [reply mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py:201]: I hope rewrite does not change the behavior of forward. We might reuse any module in mmdetection(or other repo), if we change the behavior of forward, other head which share the same forward(if it did exist) might get unexpect result.

Comment (Richard-mei) [reply mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py:201]: OK, I see what you mean, thank you.

File: mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py (hunk 4 above)
Comment (RunningLeon) [reply mmdeploy/codebase/mmdet/models/dense_heads/gfl_head.py:1]: > Hi, do I need to resubmit to add copyright:`# Copyright (c) OpenMMLab. All rights reserved.`? Also, I have a little doubt about the speed test. I have performed the speed test of the model under tensorrt on cuda and cpu respectively. Why is the FPS based on cpu very high, much higher than the speed based on cuda? Only the `--device cpu/cuda` parameter was changed during testing. My laptop cpu is r9-5900hs and gpu is 3060.**But I found that `nvidia-smi` shows gpu usage in both cases.** cpu: `python tools/test.py configs/mmdet/detection/detection_tensorrt_dynamic-320x320-1344x1344.py ../MMDetection/configs/gfl/gfl_r50_fpn_1x_coco.py --model test_work/test_GFL/end2end.engine --speed-test --device cpu` Screenshot_select-area_20220210191630 cuda: `python tools/test.py configs/mmdet/detection/detection_tensorrt_dynamic-320x320-1344x1344.py ../MMDetection/configs/gfl/gfl_r50_fpn_1x_coco.py --model test_work/test_GFL/end2end.engine --speed-test --device cuda` Screenshot_select-area_20220210191906

@Richard-mei Please add copy right in your commit. BTW, TensorRT is for cuda. Passing `cpu` to `--device` should raise error. @VVsssssk Could check if there is a bug here?

**Review by grimoire — APPROVED** [review]

**Comment by twmht** [comment]
@Richard-mei 

Have you ever tried TensorRT7 with gfl and found the accuracy drop?

**Comment by Richard-mei** [comment]
@twmht 
No, I only use tensorrt8

**Comment by Richard-mei** [comment]
@twmht 
Sorry, I just looked and I'm using tensorrt7. But don't pay attention to the precision

**Comment by grimoire** [comment]
@twmht @Richard-mei The rounding mode of IResizeLayer in TensorRT7 might give unexpect result on some shape, which might affect the accuracy. here is an example:

```python
import tensorrt as trt
import torch
import numpy as np


def main():

    input_size = [1, 1, 1, 33]

    print("create trt model")
    log_level = trt.Logger.ERROR
    logger = trt.Logger(log_level)
    builder = trt.Builder(logger)

    # build network
    EXPLICIT_BATCH = 1 << (int)(
        trt.NetworkDefinitionCreationFlag.EXPLICIT_BATCH)
    network = builder.create_network(EXPLICIT_BATCH)
    input_name = 'input'
    output_name = 'output'
    input_trt = network.add_input(name=input_name,
                                  shape=input_size,
                                  dtype=trt.float32)

    layer = network.add_resize(input_trt)
    layer.shape = tuple(input_size[:3] + [input_size[3] * 2])
    layer.resize_mode = trt.ResizeMode.NEAREST

    output = layer.get_output(0)
    output.name = output_name
    network.mark_output(output)

    # builder config
    max_workspace_size = 1 << 30
    fp16_mode = False

    builder.max_workspace_size = max_workspace_size
    builder.fp16_mode = fp16_mode

    config = builder.create_builder_config()
    config.max_workspace_size = max_workspace_size
    profile = builder.create_optimization_profile()

    # set shape
    input_shape = input_size
    profile.set_shape(input_name, input_shape, input_shape, input_shape)
    config.add_optimization_profile(profile)
    if fp16_mode:
        config.set_flag(trt.BuilderFlag.FP16)

    # build engine
    engine = builder.build_engine(network, config)
    context = engine.create_execution_context()

    print("inference")
    input_torch = torch.zeros(input_size, dtype=torch.float32).cuda().contiguous()
    input_torch[:,:,:,::2] = 1

    bindings = [None] * 2

    # set input
    idx = engine.get_binding_index(input_name)
    context.set_binding_shape(idx, tuple(input_torch.shape))
    bindings[idx] = input_torch.data_ptr()

    # set output
    idx = engine.get_binding_index(output_name)
    shape = tuple(context.get_binding_shape(idx))
    output_torch = torch.empty(shape, dtype=torch.float32).cuda()
    bindings[idx] = output_torch.data_ptr()

    context.execute_async_v2(bindings, torch.cuda.current_stream().cuda_stream)

    print("input:")
    print(input_torch.view(-1)[:20])

    print("output:")
    print(output_torch.view(-1)[:20])


if __name__ == "__main__":
    main()
```

TensorRT8 has update the layer, add `ResizeRoundMode` to set the rounding mode.

**Review by VVsssssk — APPROVED** [review]

File: mmdeploy/codebase/mmdet/models/dense_heads/__init__.py [hunk 6]
```diff
@@ -2,6 +2,7 @@
 from .base_dense_head import (base_dense_head__get_bbox,
                               base_dense_head__get_bboxes__ncnn)
 from .fovea_head import fovea_head__get_bboxes
+from .gfl_head import gfl_head__forward_single, gfl_head__get_bbox
```
Comment (grimoire) [inline mmdeploy/codebase/mmdet/models/dense_heads/__init__.py:5]: `gfl_head__forward_single` should be removed.

File: docs/en/supported_models.md [hunk 7]
```diff
@@ -14,6 +14,7 @@ The table below lists the models that are guaranteed to be exportable to other b
 | SSD[*](#note)             | MMDetection      |      Y      |    Y     |  Y   |   N   |    Y     |          [config](https://github.com/open-mmlab/mmdetection/tree/master/configs/ssd)           |
 | FoveaBox                  | MMDetection      |      Y      |    N     |  N   |   N   |    Y     |        [config](https://github.com/open-mmlab/mmdetection/tree/master/configs/foveabox)        |
 | ATSS                      | MMDetection      |      Y      |    Y     |  N   |   N   |    Y     |          [config](https://github.com/open-mmlab/mmdetection/tree/master/configs/atss)          |
+| GFL                       | MMDetection      |      Y      |    Y     |  ?   |   ?   |    ?     |          [config](https://github.com/open-mmlab/mmdetection/tree/master/configs/gfl)           |
```
Comment (RunningLeon) [inline docs/en/supported_models.md:17]: conversion of all backends should be included

**Review by RunningLeon — APPROVED** [review]
LGTM

**Comment by OpenMMLab-Assistant003** [comment]
Hi @Richard-mei！First of all, we want to express our gratitude for your significant PR in the MMDeploy project. Your contribution is highly appreciated, and we are grateful for your efforts in helping improve this open-source project during your personal time. We believe that many developers will benefit from your PR.

We would also like to invite you to join our Special Interest Group (SIG) private channel on Discord, where you can share your experiences, ideas, and build connections with like-minded peers. To join the SIG channel, simply message moderator— OpenMMLab on Discord or briefly share your open-source contributions in the #introductions channel and we will assist you. Look forward to seeing you there! Join us ：https://discord.gg/UjgXkPWNqA

If you have WeChat account，welcome to join our community on WeChat. You can add our assistant ：openmmlabwx. Please add "mmsig + Github ID" as a remark when adding friends：）
Thank you again for your contribution❤ @Richard-mei
